### PR TITLE
introduce extension runtimes

### PR DESCRIPTION
**Pull request by cecyc**
### WHY are these changes introduced?

Part of https://github.com/Shopify/ui-extensions-private/issues/1618

We need to pass the `--store` flag to the `checkout-extensions-ui-run` node process, but only for Checkout UI extensions, not for Post Purchase extensions.

### WHAT is this pull request doing?

In order to achieve the desired functionality, we have introduced various extension-specific runtimes:
* Admin
* CheckoutUiExtension
* CheckoutPostPurchase

as each one of these has very specific needs.

This PR introduces the extension-specific runtimes, and also adds the `shop` flag to extensions of type `CHECKOUT_UI_EXTENSION`.

### 🎩 Tophatting

* `cd` into a `PRODUCT_SUBSCRIPTION` extension
* run `shopify extension serve`
* observe serve behavior is unchanged

Screen Shot 2021-07-07 at 2 41 23 PM

* `cd` into a `CHECKOUT_UI_EXTENSION` or `CHECKOUT_ARGO_EXTENSION` project
* run `shopify extension serve`
* observe behavior is unchanged from before, verify `--store` flag is also being passed to the underlying node process

Screen Shot 2021-07-07 at 2 16 48 PM

* `cd` into a `CHECKOUT_POST_PURCHASE` project
* run `shopify extension serve`
* observe behavior is unchanged, verify `--store` is NOT being passed down to the underlying node process

Screen Shot 2021-07-07 at 2 17 25 PM

### Update checklist

- [ ] I've added a CHANGELOG entry for this PR (if the change is public-facing)
- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows).
- [x] I've left the version number as is (we'll handle incrementing this when releasing).

### CODE REVIEW

File: lib/project_types/extension/features/runtimes/checkout_ui_extension.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+      class CheckoutUiExtension < Default
+        IDENTIFIERS = [
+          "CHECKOUT_ARGO_EXTENSION",
+          "CHECKOUT_UI_EXTENSION",
+        ]
+
+        AVAILABLE_FLAGS = [
+          :port,
+          :public_url,
+          :shop,
```
Comment (vividviolet) [inline lib/project_types/extension/features/runtimes/checkout_ui_extension.rb:13]: Should we add the `resource_url` as well? It will be supported after this PR ships: https://github.com/Shopify/ui-extensions/pull/161

Comment (cecyc) [reply lib/project_types/extension/features/runtimes/checkout_ui_extension.rb:13]: @vividviolet I was going to do that on a separate PR :)

File: lib/project_types/extension/cli.rb [hunk 2]
```diff
@@ -62,6 +62,12 @@ module Questions
   end
 
   module Features
+    module Runtimes
+      autoload :Admin, Project.project_filepath("features/runtimes/admin")
+      autoload :Default, Project.project_filepath("features/runtimes/default")
```
Comment (joeyfreund) [inline lib/project_types/extension/cli.rb:67]: Nit: After reading the rest of the code, I feel like `Base` would be a more accurate name than `Default`. WDYT?

File: test/project_types/extension/features/argo_runtime_test.rb [hunk 3]
```diff
@@ -7,105 +7,28 @@ module Features
...
+          identifier: "PRODUCT_SUBSCRIPTION"
         )
+        assert_equal(Runtimes::Admin, runtime.class)
       end
 
-      def admin_runtime
-        ArgoRuntime.new(
-          cli: Models::NpmPackage.new(name: "@shopify/admin-ui-extensions-run", version: "0.13.0"),
-          renderer: Models::NpmPackage.new(name: "@shopify/admin-ui-extensions", version: "0.12.0")
+      def test_checkout_ui_extension_runtime_is_returned_for_checkout_ui_extensions
```
Comment (joeyfreund) [inline test/project_types/extension/features/argo_runtime_test.rb:18]: Should we add a similar test with the "legacy" identifier, `CHECKOUT_ARGO_EXTENSION`?

File: lib/project_types/extension/features/runtimes/checkout_post_purchase.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+          :port,
+          :public_url,
+        ]
+
+        def available_flags
+          AVAILABLE_FLAGS
+        end
+
+        def active_runtime?(_cli_package, identifier)
+          identifier == IDENTIFIER
```
Comment (joeyfreund) [inline lib/project_types/extension/features/runtimes/checkout_post_purchase.rb:17]: Nit: Consider looking at both, the `identifer` _and_ the `package_cli`.       

It's highly unlikely, but ... if some bug results in calling `ArgoRuntime.build` with an admin `cli_package` and a checkout `identifier`, we'll end up returning a runtime object (instead of returning `nil`).

If you make this change, I would also add a regression test (asserting that `nil` is returned when we mix an admin cli-package with a checkout identifier).

File: lib/project_types/extension/features/runtimes/checkout_ui_extension.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        ]
+
+        def available_flags
+          AVAILABLE_FLAGS
+        end
+
+        def active_runtime?(_cli_package, identifier)
+          IDENTIFIERS.include?(identifier)
```
Comment (joeyfreund) [inline lib/project_types/extension/features/runtimes/checkout_ui_extension.rb:21]: Same comment as above - Consider using both `cli_package` and `identifier`

File: test/project_types/extension/features/runtimes/admin_test.rb [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+            :name,
+            :port,
+            :public_url,
+            :renderer_version,
+            :shop,
+            :uuid,
+          ]
+
+          flags.each do |flag|
+            assert_equal runtime.supports?(flag), supported
```
Comment (joeyfreund) [inline test/project_types/extension/features/runtimes/admin_test.rb:23]: I think this should work (and will allow you to avoid defining the `supported` helper method)

```suggestion
            assert runtime.supports?(flag)
```

File: test/project_types/extension/features/runtimes/admin_test.rb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+          end
+        end
+
+        def test_unsupported_flag_is_not_supported_for_admin
+          assert_equal runtime.supports?(:fake_flag), not_supported
```
Comment (joeyfreund) [inline test/project_types/extension/features/runtimes/admin_test.rb:28]: I think this should work (and will allow you to avoid defining the `not_supported` helper method)

```suggestion
          refute runtime.supports?(:fake_flag)
```

**Review by joeyfreund — COMMENTED** [review]
I added a few minor suggestions.
Otherwise, the code looks great, but I didn't 🎩  yet.

File: test/project_types/extension/features/runtimes/admin_test.rb (hunk 7 above)
Comment (cecyc) [reply test/project_types/extension/features/runtimes/admin_test.rb:28]: same as above

**Comment by joeyfreund** [comment]
@cecyc , 🎩  with post-purchase extension isn't what I was expecting:

image

I see you just pushed a new commit, let me see if it makes a difference.     
If not, I'll ping you to see if I actually found an issue, or if it's just my local setup.

**Comment by joeyfreund** [comment]
> post purchase extension does not need the `--store` flag, so this is the expected behavior.

👍 

Also, just top-hatted checkout extension, and it looks good

image
